web/db/actions/User.ts

```diff
@@ -59,11 +72,58 @@ async function updateUserByUid(
   });
 }
 
+export interface SearchUsersParams {
+  role?: Role;
+  type?: FosterType[];
+  size?: Size[];
+  preferredBreeds?: Breed[];
+  gender?: Gender[];
+  age?: Age[];
+  temperament?: Temperament[];
+  dogsNotGoodWith?: GoodWith[];
+  medical?: Medical[];
+  behavioral?: Behavioral[];
+  houseTrained?: Status;
+  spayNeuterStatus?: Status;
+}
+
+async function searchUsers(
+  searchParams: SearchUsersParams,
+  session?: ClientSession
+): Promise<IUser[] | null> {
+  const filter: FilterQuery<IUser> = {
+    haveCompletedOnboarding: true,
+    disabled: false,
+  };
+
+  if (searchParams.role) filter.role = searchParams.role;
+  if (searchParams.type) filter.type = { $all: searchParams.type };
+  if (searchParams.size) filter.size = { $all: searchParams.size };
+  if (searchParams.preferredBreeds) filter.preferredBreeds = { $all: searchParams.preferredBreeds };
+  if (searchParams.gender) filter.gender = { $all: searchParams.gender };
+  if (searchParams.age) filter.age = { $all: searchParams.age };
+  if (searchParams.temperament) filter.temperament = { $all: searchParams.temperament };
+  if (searchParams.dogsNotGoodWith) filter.dogsNotGoodWith = { $all: searchParams.dogsNotGoodWith };
+  if (searchParams.medical) filter.medical = { $all: searchParams.medical };
+  if (searchParams.behavioral) filter.behavioral = { $all: searchParams.behavioral };
+  if (searchParams.houseTrained) filter.houseTrained = searchParams.houseTrained;
+  if (searchParams.spayNeuterStatus) filter.spayNeuterStatus = searchParams.spayNeuterStatus;
```

Comment:
Make a function that takes in `searchParams` and outputs the filter query. In this function, make a map for how each field should end up in the filter query:
```js
{
  role: searchParams.role,
  type: { $all: searchParams.preferredBreeds }, 
  ...
}
```
In the function, iterate over all the fields in searchParams (using a reduce function), adding each value of the map to the filter query object. Call this helper function from `searchUsers` to get the filter query you use.